### PR TITLE
Bitcoin wallet implementation

### PR DESCRIPTION
**Pull request by katiasuya**
.

### CODE REVIEW

File: btcwallet.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
+// Package bitcoin implemets bitcoin wallet creation
+// so that you can do some manipulations with your bitcoins
+package bitcoin
+
+import "fmt"
+
+// Btc represents a bitcoin part
+type Btc float64
+
```
Comment (papahmsolo) [inline btcwallet.go:9]: What's a bitcoin part? How about `Bitcoin` as a type name?

File: btcwallet.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+// NewWallet returns a new wallet pointer
+func NewWallet(b Btc) (w *Wallet) {
+	w = new(Wallet)
+	*w = Wallet{
+		Btc: b,
+	}
+	return
+}
```
Comment (papahmsolo) [inline btcwallet.go:23]: We can do this function more compact, you can create and return in one line.
We probably don't need a named return here, it's suggested form the name :)

File: btcwallet.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+// Display displays wallet's balance
+func (w *Wallet) Display() {
+	fmt.Printf("Your balance is %f btcs\n", w.Btc)
+}
```
Comment (papahmsolo) [inline btcwallet.go:28]: I think I should've explained it better, we need a method to _get_ the wallet's balance, we can print it to sdtout ourselves if needed.

File: btcwallet.go [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import "fmt"
+
+// Btc represents a bitcoin part
+type Btc float64
+
+// Wallet represents a bitcoin wallet
+type Wallet struct {
+	Btc
+	//should the structure have anything else or it doesn't have to be a structure at all?
+}
```
Comment (papahmsolo) [inline btcwallet.go:14]: The structure looks fine, maybe just don't inline the bitcoin type.
I think it's better to ask such questions in the PR, not it code

File: btcwallet.go [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+	if ch == 'd' || ch == 'D' {
+		(*w).Btc -= b
+		(*w2).Btc += b
+	} else if ch == 'w' || ch == 'W' {
+		(*w).Btc += b
+		(*w2).Btc -= b
+	} else {
+		//to be implemented...
+	}
+}
```
Comment (papahmsolo) [inline btcwallet.go:41]: Shouldn't we have two methods for the wallet: `Deposit` and `Withdraw`?

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Good start, but we can do better :)

File: btcwallet.go (hunk 4 above)
Comment (katiasuya) [reply btcwallet.go:14]: what do you mean by not inlining? doing bitcoin a named field?

Comment (papahmsolo) [reply btcwallet.go]: Yep

File: btcwallet.go [hunk 6]
```diff
@@ -1,41 +1,47 @@
...
-	return
 }
 
-// Display displays wallet's balance
-func (w *Wallet) Display() {
-	fmt.Printf("Your balance is %f btcs\n", w.Btc)
+// Display returns wallet's balance
+func (w *Wallet) Display() Bitcoin {
+	return w.Blnc
 }
```
Comment (papahmsolo) [inline btcwallet.go]: Let's name it `Balance`

File: btcwallet.go [hunk 7]
```diff
@@ -1,41 +1,47 @@
...
-		(*w2).Btc -= b
+// Deposit deposits(adds) bitcoins
+func (w *Wallet) Deposit(d Bitcoin) {
+	if d >= 0 {
+		(*w).Blnc += d
 	} else {
-		//to be implemented...
+		fmt.Println("Invalid deposit value")
+	}
+}
```
Comment (papahmsolo) [inline btcwallet.go]: - Do we need `else` here? Early return helps here
- Is this `Println` good here?

File: btcwallet.go [hunk 8]
```diff
@@ -1,41 +1,47 @@
...
+	if d < 0 {
+		fmt.Println("Invalid withdraw value")
+	} else {
+		if d > w.Blnc {
+			fmt.Println("Insufficient funds")
+		} else {
+			(*w).Blnc -= d
+		}
 	}
 }
```
Comment (papahmsolo) [inline btcwallet.go]: Again, early return helps. Have a look: https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88

File: btcwallet_test.go [hunk 9]
```diff
@@ -1,63 +1,143 @@
...
+	"testing"
+)
 
-// just for me: !!!!! get rid of float64 inaccuracy !!!!!!!
+// TestDisplay tests Display function
+func TestDisplay(t *testing.T) {
 
-// TestDepWith tests DepWith function
-func TestDepWith(t *testing.T) {
+	w := NewWallet(0)
```
Comment (papahmsolo) [inline btcwallet_test.go:11]: Even though float and Bitcoin compatible, what does `NewWallet` take as an argument?

**Review by papahmsolo — CHANGES_REQUESTED** [review]
See comments

File: btcwallet.go (hunk 6 above)
Comment (katiasuya) [reply btcwallet.go]: I was looking at some gophers's repositories and I saw it common to have blnc instead of balance.

**Comment by katiasuya** [comment]
Have I done it right by just changing balance value instead of creating a new wallet every time in the for loop of Display, Deposit and Withdraw functions?
Should I create test and example functions for NewWallet?

File: btcwallet.go [hunk 10]
```diff
@@ -1,47 +1,46 @@
...
+)
 
 // Bitcoin represents a bitcoin
 type Bitcoin float64
 
 // Wallet represents a bitcoin wallet
 type Wallet struct {
-	Blnc Bitcoin
+	Balance Bitcoin
 }
```
Comment (papahmsolo) [inline btcwallet.go]: Why would we want to make it exported or unexported? I think can make the right choice when you'll answer this question.

File: btcwallet.go [hunk 11]
```diff
@@ -1,47 +1,46 @@
...
-func (w *Wallet) Deposit(d Bitcoin) {
-	if d >= 0 {
-		(*w).Blnc += d
-	} else {
-		fmt.Println("Invalid deposit value")
+func (w *Wallet) Deposit(b Bitcoin) error {
+	if b < 0 {
+		return errors.New("Invalid deposit value")
 	}
+	(*w).Balance += b
```
Comment (papahmsolo) [inline btcwallet.go:32]: The pointer will be dereferenced automatically for you, no need for parentheses.

File: btcwallet.go [hunk 12]
```diff
@@ -1,47 +1,46 @@
...
 
 // Deposit deposits(adds) bitcoins
-func (w *Wallet) Deposit(d Bitcoin) {
-	if d >= 0 {
-		(*w).Blnc += d
-	} else {
-		fmt.Println("Invalid deposit value")
+func (w *Wallet) Deposit(b Bitcoin) error {
+	if b < 0 {
+		return errors.New("Invalid deposit value")
```
Comment (papahmsolo) [inline btcwallet.go:30]: - https://github.com/golang/go/wiki/CodeReviewComments#error-strings
- it's always good to indicate why an error occurs, such as why the deposit value is invalid?

File: btcwallet_test.go [hunk 13]
```diff
@@ -124,20 +129,29 @@ func TestWithdraw(t *testing.T) {
...
-// ExampleWithdraw provides working and runnable function execution example
+// ExampleWithdraw provides Withdraw function example
 func ExampleWithdraw() {
-	w := NewWallet(43.765)
-	w.Withdraw(1.235)
-	fmt.Println(w.Blnc)
-	// Output: 42.53
+	w := NewWallet()
+	_ = w.Deposit(1.235)
+	err := w.Withdraw(1.235)
```
Comment (papahmsolo) [inline btcwallet_test.go:149]: Does our wallet work with `float` or `Bitcoin` type?

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Yeah, looks good!
I don't think we need to test the new wallet here.

Also, read up on `t.Run()` and see how can we leverage it for wallet testing.

File: btcwallet.go (hunk 10 above)
Comment (katiasuya) [reply btcwallet.go]: I just want it to be accessible for other packages. Or we just don't need it?

Comment (papahmsolo) [reply btcwallet.go]: What's the problem with having it exported and why do we have `Deposit()` and `Withdraw`?

File: btcwallet.go [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+
+// NewWallet returns an empty wallet pointer
+func NewWallet() (w *Wallet) {
+	return &Wallet{}
+}
+
+// Display returns wallet's balance
+func (w *Wallet) Display() Bitcoin {
+	return w.Balance
+}
```
Comment (papahmsolo) [inline btcwallet.go:25]: Let's call this method `Balance()` as it's returning the wallet's balance.

File: btcwallet_test.go [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+	}
+}
+
+// ExampleDisplay provides a Display function example
+func ExampleDisplay() {
+	w := NewWallet()
+	_ = w.Deposit(Bitcoin(43.765))
+	fmt.Println(w.Display())
+	// Output: 43.765
+}
```
Comment (papahmsolo) [inline btcwallet_test.go:46]: It's a good idea to handle errors like in the previous exmples.

File: btcwallet_test.go [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+
+	tests := []struct {
+		bal Bitcoin
+		dep Bitcoin
+		exp Bitcoin
+	}{
+		{
+			bal: 100,
+			dep: 5,
+			exp: 105,
```
Comment (papahmsolo) [inline btcwallet_test.go:60]: Aren't these also `Bitcoins`?

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Looks good! Let's make it better 👍 

Can anything go wrong with the `Display()`? I'd use it to assert the balance after we do some manipulations with the wallet. The ExampleDisplay looks good though

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Getting better!)

Also, see my previous comments about `t.Run()` and `Bitcoin` type in test data 👍

File: btcwallet_test.go [hunk 17]
```diff
@@ -0,0 +1,158 @@
...
+	if err != nil {
+		fmt.Println(err)
+		return
+	}
+
+	err2 := w.Withdraw(Bitcoin(0.135))
+	if err2 != nil {
+		fmt.Println(err2)
+		return
+	}
```
Comment (papahmsolo) [inline btcwallet_test.go:154]: We can reuse `err` here

File: btcwallet_test.go [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+				return
+			}
+
+			result := w.Balance()
+			if result != tt.exp {
+				t.Errorf("Expected %g, got %g", tt.exp, result)
+			}
+		})
+	}
+}
```
Comment (papahmsolo) [inline btcwallet_test.go:72]: Maybe we can do something with these errors? LIke, check if we got the right one.

**Review by papahmsolo — CHANGES_REQUESTED** [review]
This looks great!

Let's improve our tests a little bit, and test different _scenarios_ with deposit and withdrawal.

File: btcwallet.go [hunk 19]
```diff
@@ -6,6 +6,15 @@ import (
...
 )
 
+// ErrNegativeDep is a negative deposit error
+var ErrNegativeDep = errors.New("negative deposit value")
+
+// ErrNegativeWith is a negative withdraw error
+var ErrNegativeWith = errors.New("negative withdraw value")
+
+// ErrInsufFunds is a insufficient funds error
+var ErrInsufFunds = errors.New("insufficient funds")
```
Comment (papahmsolo) [inline btcwallet.go:16]: Nice! Let's group them together. Also, you can spell the whole words in the error name just for clarity

File: btcwallet_test.go [hunk 20]
```diff
@@ -11,54 +11,52 @@ func TestWallet(t *testing.T) {
...
 				return
 			}
 
 			err = w.Withdraw(tt.with)
-			if err != nil {
+			if err == ErrNegativeWith {
+				fmt.Println(err)
+				return
+			}
+			if err == ErrInsufFunds {
```
Comment (papahmsolo) [inline btcwallet_test.go]: Suppose we're making a perfectly legal withdrawal, say we have 10 bitcoins on our balance and want to withdraw 5. And for some reason, our `Withdraw` method returns an error `ErrInsufFunds` this test will pass, but we don't want that, do we?

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Looks good!

Let's just assert errors and we're ready to go.

File: btcwallet_test.go [hunk 21]
```diff
@@ -41,50 +40,99 @@ func TestWallet(t *testing.T) {
...
-			if result != tt.exp {
-				t.Errorf("Expected %g, got %g", tt.exp, result)
-			}
-		})
+	t.Run(tests[0].name, func(t *testing.T) {
+
+		w := NewWallet()
+
+		_ = w.Deposit(tests[0].dep)
+		_ = w.Withdraw(tests[0].with)
```
Comment (papahmsolo) [inline btcwallet_test.go:48]: Almost never a good idea to ignore errors :)

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Looks good! Let's make it better :)

File: btcwallet_test.go [hunk 22]
```diff
@@ -0,0 +1,162 @@
...
+)
+
+// TestWallet tests Deposit, Withdraw and Balance functions.
+func TestWallet(t *testing.T) {
+
+	type test struct {
+		dep  Bitcoin
+		with Bitcoin
+		exp  Bitcoin
+	}
```
Comment (papahmsolo) [inline btcwallet_test.go:15]: Do actually we need this `test` type? We can get by with `Bitcoin` variables inside each `t.Run()`

File: btcwallet_test.go [hunk 23]
```diff
@@ -0,0 +1,162 @@
...
+		if err != nil {
+			t.Errorf("No error expected, got %v error", err)
+			return
+		}
+
+		err = w.Withdraw(tt.with)
+		if err != nil {
+			t.Errorf("No error expected, got %v error", err)
+			return
+		}
```
Comment (papahmsolo) [inline btcwallet_test.go:36]: This feels like repeating yourself, doesn't it? Maybe we can create a helper function to assert the error? We can reuse it in all tests! Also, it's handy to mark such functions with `t.Helper()`

File: btcwallet_test.go [hunk 24]
```diff
@@ -0,0 +1,162 @@
...
+		err = w.Deposit(tt.dep)
+		if err != ErrNegativeDeposit {
+			t.Errorf("Expected %v, got %v error", ErrNegativeDeposit, err)
+			return
+		}
+
+		if w.Balance() != tt.exp {
+			t.Errorf("Expected %g, got %g", tt.exp, w.Balance())
+			return
+		}
```
Comment (papahmsolo) [inline btcwallet_test.go:68]: Same here, we're asserting balance many times, worth creating a func.

**Review by papahmsolo — CHANGES_REQUESTED** [review]
This looks excellent! Great job!

File: btcwallet_test.go [hunk 25]
```diff
@@ -5,133 +5,95 @@ import (
...
 )
 
-// TestWallet tests Deposit, Withdraw and Balance functions.
-func TestWallet(t *testing.T) {
+func assertNoError(t *testing.T, err error) {
+	t.Helper()
+	if err != nil {
+		t.Errorf("No error expected, got %v error", err)
+		return
+	}
```
Comment (papahmsolo) [inline btcwallet_test.go:13]: No need to return here, it'll return anyway

File: btcwallet_test.go [hunk 26]
```diff
@@ -5,133 +5,95 @@ import (
...
 	})
 }
 
+func assertErrorInExample(err error, assertErr error) {
+	if err == assertErr {
+		fmt.Println(err)
+		return
+	}
+}
+
```
Comment (papahmsolo) [inline btcwallet_test.go:90]: I don't think we need to assert errors in Examples. They are intended for showing users how to use the wallet.
Something like this should be just fine:
```
	w := NewWallet()
	w.Deposit(Bitcoin(1.235))
	fmt.Println(w.Balance())
```

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Brilliant! Tests are so clean now.

Some small fixes and we're ready to go.

**Review by papahmsolo — APPROVED** [review]

**Pull request merged**